File: mobile/src/api/social.ts
import { api } from './client';
import type {
  ChallengeItem,
  ChallengesList,
  ConversationItem,
  ConversationThread,
  FriendshipItem,
  MessageItem,
  PendingRequests,
  PublicProfile,
  UserCard,
} from './types';

// src/api/social.ts — web'deki lib/api.ts içindeki socialApi'nin mobil
// karşılığı, birebir aynı endpoint'ler. Meydan okuma (challenge) metodları
// forward-compat için eklendi ama mobil Faz 3 UI'ı henüz bunları çağırmıyor
// (bkz. friends.tsx üstündeki not).
export const socialApi = {
  searchUsers: async (q: string, limit = 20): Promise<UserCard[]> => {
    const res = await api.get('/social/users/search', { params: { q, limit } });
    return res.data.items;
  },

  // ── Arkadaşlık ──
  getFriends: async (): Promise<FriendshipItem[]> => {
    const res = await api.get('/social/friends');
    return res.data.items;
  },
  getPendingRequests: async (): Promise<PendingRequests> => {
    const res = await api.get<PendingRequests>('/social/friends/pending');
    return res.data;
  },
  sendFriendRequest: async (username: string): Promise<FriendshipItem> => {
    const res = await api.post<FriendshipItem>('/social/friends/request', { username });
    return res.data;
  },
  acceptFriendRequest: async (friendshipId: string): Promise<FriendshipItem> => {
    const res = await api.post<FriendshipItem>(`/social/friends/${friendshipId}/accept`);
    return res.data;
  },
  declineFriendRequest: async (friendshipId: string): Promise<FriendshipItem> => {
    const res = await api.post<FriendshipItem>(`/social/friends/${friendshipId}/decline`);
    return res.data;
  },
  removeFriend: async (userId: string): Promise<void> => {
    await api.delete(`/social/friends/${userId}`);
  },

  // ── Takip ──
  follow: async (userId: string): Promise<void> => {
    await api.post(`/social/follow/${userId}`);
  },
  unfollow: async (userId: string): Promise<void> => {
    await api.delete(`/social/follow/${userId}`);
  },
  getFollowers: async (): Promise<UserCard[]> => {
    const res = await api.get('/social/followers');
    return res.data.items;
  },
  getFollowing: async (): Promise<UserCard[]> => {
    const res = await api.get('/social/following');
    return res.data.items;
  },

  // ── Herkese açık profil ──
  getPublicProfile: async (username: string): Promise<PublicProfile> => {
    const res = await api.get<PublicProfile>(`/social/profile/${username}`);
    return res.data;
  },

  // ── Engelleme ──
  blockUser: async (userId: string): Promise<void> => {
    await api.post(`/social/block/${userId}`);
  },
  unblockUser: async (userId: string): Promise<void> => {
    await api.delete(`/social/block/${userId}`);
  },
  getBlockedUsers: async (): Promise<UserCard[]> => {
    const res = await api.get('/social/blocked');
    return res.data.items;
  },

  // ── Şikayet/Rapor — Guideline 1.2 (Safety - UGC) için engellemeye ek
  // moderasyon yolu. Kullanıcı ve/veya belirli bir mesajı raporlar.
  reportUser: async (userId: string, reason: string, messageId?: string): Promise<void> => {
    await api.post(`/social/report/${userId}`, { reason, message_id: messageId });
  },

  // ── Mesajlaşma — polling tabanlı, gerçek zamanlı değil (web ile aynı) ──
  getConversations: async (): Promise<ConversationItem[]> => {
    const res = await api.get('/social/conversations');
    return res.data.items;
  },
  getConversationThread: async (username: string): Promise<ConversationThread> => {
    const res = await api.get<ConversationThread>(`/social/conversations/${username}`);
    return res.data;
  },
  sendMessage: async (username: string, body: string): Promise<MessageItem> => {
    const res = await api.post<MessageItem>(`/social/conversations/${username}`, { body });
    return res.data;
  },
  getUnreadMessageCount: async (): Promise<number> => {
    const res = await api.get('/social/messages/unread-count');
    return res.data.unread_count;
  },

  // ── Meydan okuma — forward-compat, mobil UI'da henüz kullanılmıyor ──
  createChallenge: async (username: string, mode: string): Promise<ChallengeItem> => {
    const res = await api.post<ChallengeItem>('/social/challenges', { username, mode });
    return res.data;
  },
  getChallenges: async (): Promise<ChallengesList> => {
    const res = await api.get<ChallengesList>('/social/challenges');
    return res.data;
  },
  acceptChallenge: async (challengeId: string): Promise<ChallengeItem> => {
    const res = await api.post<ChallengeItem>(`/social/challenges/${challengeId}/accept`);
    return res.data;
  },
  declineChallenge: async (challengeId: string): Promise<ChallengeItem> => {
    const res = await api.post<ChallengeItem>(`/social/challenges/${challengeId}/decline`);
    return res.data;
  },
  cancelChallenge: async (challengeId: string): Promise<void> => {
    await api.post(`/social/challenges/${challengeId}/cancel`);
  },
  submitChallengeScore: async (challengeId: string, sessionId: string): Promise<ChallengeItem> => {
    const res = await api.post<ChallengeItem>(`/social/challenges/${challengeId}/submit`, {
      session_id: sessionId,
    });
    return res.data;
  },
};

File: backend/app/api/routes/social.py
"""
backend/app/api/routes/social.py

Madde 6, Faz 1 — Arkadaşlık + Takip + Profil görüntüleme.
Madde 6, Faz 2 — Engelleme + Mesajlaşma.
Madde 6, Faz 3 — Meydan okuma.
Backend: /api/v1/social/*
"""

from fastapi import APIRouter, Depends

from app.core.auth import get_current_user
from app.schemas.social import (
    BlockedListResponse,
    ChallengeCreate,
    ChallengeItem,
    ChallengesListResponse,
    ChallengeSubmitScore,
    ConversationsListResponse,
    ConversationThreadResponse,
    FollowListResponse,
    FriendRequestCreate,
    FriendshipItem,
    FriendsListResponse,
    MessageCreate,
    MessageItem,
    PendingRequestsResponse,
    PublicProfileResponse,
    ReportCreate,
    UnreadCountResponse,
    UserSearchResponse,
)
from app.services import (
    block_service,
    challenge_service,
    follow_service,
    friends_service,
    messaging_service,
    public_profile_service,
    report_service,
)

router = APIRouter()


# ── Kullanıcı arama ──────────────────────────────────────────────────
@router.get("/users/search", response_model=UserSearchResponse)
async def search_users(q: str = "", limit: int = 20, current_user=Depends(get_current_user)):
    items = friends_service.search_users(current_user.id, q, limit)
    return {"items": items}


# ── Arkadaşlık ───────────────────────────────────────────────────────
@router.get("/friends", response_model=FriendsListResponse)
async def get_friends(current_user=Depends(get_current_user)):
    return {"items": friends_service.list_friends(current_user.id)}


@router.get("/friends/pending", response_model=PendingRequestsResponse)
async def get_pending_requests(current_user=Depends(get_current_user)):
    return friends_service.list_pending(current_user.id)


@router.post("/friends/request", response_model=FriendshipItem)
async def send_friend_request(data: FriendRequestCreate, current_user=Depends(get_current_user)):
    return friends_service.send_friend_request(current_user.id, data.username)


@router.post("/friends/{friendship_id}/accept", response_model=FriendshipItem)
async def accept_friend_request(friendship_id: str, current_user=Depends(get_current_user)):
    return friends_service.respond_to_request(current_user.id, friendship_id, accept=True)


@router.post("/friends/{friendship_id}/decline", response_model=FriendshipItem)
async def decline_friend_request(friendship_id: str, current_user=Depends(get_current_user)):
    return friends_service.respond_to_request(current_user.id, friendship_id, accept=False)


@router.delete("/friends/{user_id}")
async def remove_friend(user_id: str, current_user=Depends(get_current_user)):
    friends_service.remove_friend(current_user.id, user_id)
    return {"message": "ok"}


# ── Takip ────────────────────────────────────────────────────────────
@router.post("/follow/{user_id}")
async def follow_user(user_id: str, current_user=Depends(get_current_user)):
    follow_service.follow_user(current_user.id, user_id)
    return {"message": "ok"}


@router.delete("/follow/{user_id}")
async def unfollow_user(user_id: str, current_user=Depends(get_current_user)):
    follow_service.unfollow_user(current_user.id, user_id)
    return {"message": "ok"}


@router.get("/followers", response_model=FollowListResponse)
async def get_followers(current_user=Depends(get_current_user)):
    items = follow_service.list_followers(current_user.id)
    return {"items": items, "total": len(items)}


@router.get("/following", response_model=FollowListResponse)
async def get_following(current_user=Depends(get_current_user)):
    items = follow_service.list_following(current_user.id)
    return {"items": items, "total": len(items)}


# ── Herkese açık profil ─────────────────────────────────────────────
@router.get("/profile/{username}", response_model=PublicProfileResponse)
async def get_public_profile(username: str, current_user=Depends(get_current_user)):
    return public_profile_service.get_public_profile(username, current_user.id)


# ── Engelleme (Faz 2) ────────────────────────────────────────────────
@router.post("/block/{user_id}")
async def block_user(user_id: str, current_user=Depends(get_current_user)):
    block_service.block_user(current_user.id, user_id)
    return {"message": "ok"}


@router.delete("/block/{user_id}")
async def unblock_user(user_id: str, current_user=Depends(get_current_user)):
    block_service.unblock_user(current_user.id, user_id)
    return {"message": "ok"}


@router.get("/blocked", response_model=BlockedListResponse)
async def get_blocked_users(current_user=Depends(get_current_user)):
    return {"items": block_service.list_blocked(current_user.id)}


# ── Şikayet/Rapor — engellemeye ek moderasyon yolu, bkz. Guideline 1.2 ─
@router.post("/report/{user_id}")
async def report_user(user_id: str, data: ReportCreate, current_user=Depends(get_current_user)):
    report_service.create_report(current_user.id, user_id, data.reason, data.details, data.message_id)
    return {"message": "ok"}


# ── Mesajlaşma (Faz 2) — polling tabanlı, bkz. messaging_service.py ──
@router.get("/conversations", response_model=ConversationsListResponse)
async def get_conversations(current_user=Depends(get_current_user)):
    return {"items": messaging_service.list_conversations(current_user.id)}


@router.get("/conversations/{username}", response_model=ConversationThreadResponse)
async def get_conversation_thread(username: str, current_user=Depends(get_current_user)):
    return messaging_service.get_thread(current_user.id, username)


@router.post("/conversations/{username}", response_model=MessageItem)
async def post_message(username: str, data: MessageCreate, current_user=Depends(get_current_user)):
    return messaging_service.send_message(current_user.id, username, data.body)


@router.get("/messages/unread-count", response_model=UnreadCountResponse)
async def get_unread_message_count(current_user=Depends(get_current_user)):
    return {"unread_count": messaging_service.unread_total(current_user.id)}


# ── Meydan okuma (Faz 3) — bkz. challenge_service.py ─────────────────
@router.post("/challenges", response_model=ChallengeItem)
async def create_challenge(data: ChallengeCreate, current_user=Depends(get_current_user)):
    return challenge_service.create_challenge(current_user.id, data.username, data.mode)


@router.get("/challenges", response_model=ChallengesListResponse)
async def get_challenges(current_user=Depends(get_current_user)):
    return challenge_service.list_challenges(current_user.id)


@router.post("/challenges/{challenge_id}/accept", response_model=ChallengeItem)
async def accept_challenge(challenge_id: str, current_user=Depends(get_current_user)):
    return challenge_service.respond_to_challenge(current_user.id, challenge_id, accept=True)


@router.post("/challenges/{challenge_id}/decline", response_model=ChallengeItem)
async def decline_challenge(challenge_id: str, current_user=Depends(get_current_user)):
    return challenge_service.respond_to_challenge(current_user.id, challenge_id, accept=False)


@router.post("/challenges/{challenge_id}/cancel")
async def cancel_challenge(challenge_id: str, current_user=Depends(get_current_user)):
    challenge_service.cancel_challenge(current_user.id, challenge_id)
    return {"message": "ok"}


@router.post("/challenges/{challenge_id}/submit", response_model=ChallengeItem)
async def submit_challenge_score(
    challenge_id: str, data: ChallengeSubmitScore, current_user=Depends(get_current_user)
):
    return challenge_service.submit_score(current_user.id, challenge_id, data.session_id)
